packages/@sfx/core/src/utils/core.ts

```diff
@@ -0,0 +1,79 @@
+import { Plugin, PluginRegistry } from '../plugin';
+
+/**
+ * Calculates the missing dependencies of the given plugins. The given
+ * plugins and the plugins in the registry are eligible to satisfy
+ * dependencies.
+ *
+ * @param plugins The plugins whose dependencies should be checked.
+ * @param registry The plugin registry containing all registered plugins.
+ * @returns An array of names of missing plugins.
+ */
+export function calculateMissingDependencies(plugins: Plugin[], registry: object): string[] {
+  const available = [
+    ...Object.keys(registry),
+    ...plugins.map(({ metadata: { name }}) => name),
+  ];
+  const availableSet = new Set(available);
+
+  const required = plugins.reduce((memo, plugin) => {
```

Comment:
`reduce` seems like the wrong method to use on the array here as the return value isn't always being reduced in size compared to the inputs (ignoring overall structure of the plugins array). While both methods can technically work, wouldn't `map` be the more correct method?

Reply:
`reduce()` is being used because each plugin in the array actually contains an array of dependencies. This method is essentially just flattening those arrays. There may be a way to use `Array.prototype.flat()` in combination with `map()`, here, but I don't know that it will be any simpler.

Reply:
`reduce` is correct here, though I admit that there are more performant (though possibly less readable) ways of achieving this. The trick is to realize that `plugin.metadata.depends` is an array of strings and not simply a string, so this `reduce` invocation actually flattens the arrays into a single array. `map` wouldn't be able to do that without being misused.

Reply:
Yup, I missed that part when I read through the code. Thanks for the clarification.